README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# blocklength <img src="man/figures/logo.svg" alt="" style="padding-left: 20px" align="right" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/Alec-Stashevsky/blocklength/workflows/R-CMD-check/badge.svg)](https://github.com/Alec-Stashevsky/blocklength/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/blocklength)](https://CRAN.R-project.org/package=blocklength)
[![Downloads](https://cranlogs.r-pkg.org/badges/grand-total/blocklength?color=brightgreen)](https://CRAN.R-project.org/package=blocklength)
[![CRAN/METACRAN](https://img.shields.io/cran/l/blocklength)](https://CRAN.R-project.org/package=blocklength)
[![codecov](https://codecov.io/gh/Alec-Stashevsky/blocklength/branch/main/graph/badge.svg?token=U2RFAU594R)](https://app.codecov.io/gh/Alec-Stashevsky/blocklength)
<!-- badges: end -->

`blocklength` is an R package used to automatically select the
block-length parameter for a block-bootstrap. It is meant for use with
dependent data such as stationary time series.

## The Story

Regular bootstrap methods rely on assumptions that observations are
independent and identically distributed (*i.i.d.*), but this assumption
fails for many types of time series because we would expect the
observation in the previous period to have some explanatory power over
the current observation. This could occur in any time series from
unemployment rates, stock prices, biological data, etc. A time series
that is *i.i.d.* would look like white noise, since the following
observation would be totally independent of the previous one (random).

To get around this problem, we can retain some of this *time-dependence*
by breaking-up a time series into a number of blocks with length *l*.
Instead of sampling each observation randomly (with replacement) like a
regular bootstrap, we can resample these *blocks* at random. This way
within each block the time-dependence is preserved.

The problem with the block bootstrap is the high sensitivity to the
choice of block-length, or the number of blocks to break the time series
into.

The goal of `blocklength` is to simplify and automate the process of
selecting a block-length to perform a bootstrap on dependent data.
`blocklength` has several functions that take their name from the
authors who have proposed them. Currently, there are two methods
available:

1.  `hhj()` takes its name from the [Hall, Horowitz, and
    Jing (1995)](https://academic.oup.com/biomet/article-abstract/82/3/561/260651?redirectedFrom=fulltext)
    “HHJ” method to select the optimal block-length using a
    cross-validation algorithm which minimizes the mean squared error
    *(MSE)* incurred by the bootstrap at various block-lengths.

2.  `pwsd()` takes its name from the [Politis and
    White (2004)](https://doi.org/10.1081/ETC-120028836) Spectral
    Density “PWSD” Plug-in method to automatically select the optimal
    block-length using spectral density estimation via “flat-top” lag
    windows of [Politis and Romano
    (1995).](https://doi.org/10.1111/j.1467-9892.1995.tb00223.x)

Under the hood, `hhj()` uses the moving block bootstrap (MBB) procedure
according to [Künsch
(1989)](https://projecteuclid.org/euclid.aos/1176347265) which resamples
blocks from a set of overlapping sub-samples with a fixed block-length.
However, the results of `hhj()` may be generalized to other block
bootstrap procedures such as the *stationary bootstrap* of [Politis and
Romano
(1994).](https://www.tandfonline.com/doi/abs/10.1080/01621459.1994.10476870)

Compared to `pwsd()`, `hhj()` is more computationally intensive as it
relies on iterative resampling processes that optimize the MSE function
over each possible block-length (or a select grid of block-lengths),
while `pwsd()` is a simpler “plug-in” rule that uses auto-correlations,
auto-covariance, and the spectral density of the series to optimize the
choice of block-length.

## Installation

You can install the released version from
[CRAN](https://cran.r-project.org/package=blocklength) with:

``` r
install.packages("blocklength")
```

You can install the development version from
[GitHub](https://github.com/Alec-Stashevsky/blocklength) with:

``` r
# install.packages("devtools")
devtools::install_github("Alec-Stashevsky/blocklength")
```

## Upcoming changes

-   change parallel to `{foreach}` or `{future}` (pending user
    feedback - let me know!)

## Use Case

We want to select the optimal block-length to perform a block bootstrap
on a simulated autoregressive *AR(1)* time series.

First we will generate the time series:

``` r
library(blocklength)

# Simulate AR(1) time series
series <- stats::arima.sim(model = list(order = c(1, 0, 0), ar = 0.5),
                           n = 500, rand.gen = rnorm)
```

Now, we can find the optimal block-length to perform a block-bootstrap.
We do this using two methods.

### 1. The Hall, Horowitz, and Jing (1995) “HHJ” Method

``` r
## Using the HHJ Algorithm with overlapping subsamples of width 10
hhj(series, sub_sample = 10, k = "bias/variance")
#>  Pilot block length is: 3
#> Registered S3 method overwritten by 'quantmod':
#>   method            from
#>   as.zoo.data.frame zoo
#> Performing minimization may take some time
#> Calculating MSE for each level in subsample: 10 function evaluations required.
#>  Chosen block length: 11  After iteration: 1
```

<img src="man/figures/README-hhj-1.svg" width="100%" style="display: block; margin: auto;" />

    #>  Converged at block length (l): 11

<img src="man/figures/README-hhj-2.svg" width="100%" style="display: block; margin: auto;" />

    #> $`Optimal Block Length`
    #> [1] 11
    #> 
    #> $`Subsample block size (m)`
    #> [1] 10
    #> 
    #> $`MSE Data`
    #>    Iteration BlockLength       MSE
    #> 1          1           4 0.5428141
    #> 2          1           7 0.5159969
    #> 3          1          11 0.5058036
    #> 4          1          15 0.5101443
    #> 5          1          18 0.5173569
    #> 6          1          22 0.5285706
    #> 7          1          26 0.5307042
    #> 8          1          29 0.5459579
    #> 9          1          33 0.5596887
    #> 10         1          37 0.5536393
    #> 11         2           4 0.5404021
    #> 12         2           7 0.5165052
    #> 13         2          11 0.5083257
    #> 14         2          15 0.5096941
    #> 15         2          18 0.5099022
    #> 16         2          22 0.5251345
    #> 17         2          26 0.5297922
    #> 18         2          29 0.5411978
    #> 19         2          33 0.5595094
    #> 20         2          37 0.5541540
    #> 
    #> $Iterations
    #> [1] 2
    #> 
    #> $Series
    #> [1] "series"
    #> 
    #> $Call
    #> hhj(series = series, sub_sample = 10, k = "bias/variance")
    #> 
    #> attr(,"class")
    #> [1] "hhj"

### 2. The Politis and White (2004) Spectral Density Estimation “PWSD” Method

``` r
# Coerce time series to data.frame
data <- data.frame("AR1" = series)

# Using Politis and White (2004) Spectral Density Estimation
pwsd(data) 
```

<img src="man/figures/README-pwsd-1.svg" width="100%" style="display: block; margin: auto;" />

    #> $BlockLength
    #>     b_Stationary b_Circular
    #> AR1     9.327923   10.67781
    #> 
    #> $Acf
    #> $Acf$AR1
    #> 
    #> Autocorrelations of series 'data[, i]', by lag
    #> 
    #>      0      1      2      3      4      5      6      7      8      9     10 
    #>  1.000  0.550  0.291  0.175  0.107  0.066  0.049  0.010 -0.078 -0.057 -0.071 
    #>     11     12     13     14     15     16     17     18     19     20     21 
    #> -0.081 -0.071 -0.097 -0.073 -0.054 -0.010  0.014 -0.030 -0.042 -0.020 -0.073 
    #>     22     23     24     25     26     27     28 
    #> -0.088 -0.050 -0.058 -0.097 -0.049  0.021  0.078 
    #> 
    #> 
    #> $parameters
    #>        n k        c K_N M_max b_max m_hat M rho_k_critical
    #> [1,] 500 1 1.959964   5    28    68     3 6      0.1439999
    #> 
    #> $Call
    #> pwsd(data = data)
    #> 
    #> attr(,"class")
    #> [1] "pwsd"

We can see that both methods produce similar results for a block-length
of 9 or 11 depending on the type of bootstrap method used.

## Acknowledgements

A big shoutout to Malina Cheeneebash for designing the `blocklength` hex
sticker! Also to Sergio Armella and [Simon P.
Couch](https://www.simonpcouch.com) for their help and feedback!


NEWS.md
# blocklength 0.1.5

* Minor adjustments to documentation per CRAN requests.


# blocklength 0.1.4


## Bug Fixes

* In `pwsd()`, `m_hat` was indexing the first insignificant lag of the correlation structure rather than the first *significant* lag prior to the consecutive run of insignificant lags.

* `pwsd()` and `plot.pwsd()` now output significance bands that match `rho_k_critical` exactly. Prior to this, the significance bands on the correlogram output were generated using the `plot.acf(ci = )` argument which led to misleading graphical representations of the implied hypothesis test's critical value in the PWSD method.

* In `hhj()`, `sub_block_length` has been changed to `sub_sample` to avoid confusion with the other tuning parameter, `pilot_block_length`


## Minor Changes

* A new vignette has been included on tuning and diagnosing problematic output from the selection functions!

* `pwsd()` now includes a new argument to override the implied hypothesis test by setting `m_hat` directly.

* In `pwsd()`, `rho.k.critical` has been changed to snake case `rho_k_critical` in the `$parameters` matrix from output of class 'pwsd' objects from `pwsd().`

* In `hhj()`, if `subsample = ` is set directly it is now rounded to the nearest whole number.

* `plot.pwsd()` now includes "darkmagenta" significance lines to match `pwsd().`

* `plot.pwsd()` now explicitly includes an option to customize title with `main = .`

* `hhj()` now includes a warning message if the supplied iteration limit `n_iter` is reached but still outputs an object of class 'hhj.'


# blocklength 0.1.3

* This is the first submission accepted to CRAN
* Minor changes to documentation and testing of examples


# blocklength 0.1.0

* This is the first submission of blocklength package.


tests/testthat.R
library(testthat)
library(blocklength)

test_check("blocklength")


tests/testthat/test-hhj.R
test_that("hhjMSE function works", {

  series <- stats::arima.sim(list(order = c(1, 0, 0), ar = 0.5),
    n = 100, innov = rnorm(100))

  n <- length(series)

  m <- round(n^(1 / 5) * n^(1 / 3))

  # Initialize overlapping sub samples list
  series.list <- vector(mode = "list", length = length((n - m + 1)))

  for (i in seq_len(length.out = (n - m + 1))) {
    series.list[[i]] <- series[seq(from = i, to = (i + m - 1), by = 1)]
  }

  # Tests
  expect_gte(
    hhjMSE(1, n = n, m = m,
      series.list = series.list, nb = 5,
      bofb = 1L, v_star = 2),
    0)

  expect_error(
    hhjMSE(0, n = n, m = m,
      series.list = series.list, nb = 5,
      bofb = 1L, v_star = 2)
    )

  expect_error(
    hhjMSE(-1, n = n, m = m,
      series.list = series.list, nb = 5,
      bofb = 1L, v_star = 2)
  )

  expect_equal(
    hhjMSE(1, n = n, m = m,
      series.list = series, nb = 5,
      bofb = 1L, v_star = 2),
    NA_integer_)

})


test_that("hhj function works", {

  series <- stats::arima.sim(list(order = c(1, 0, 0), ar = 0.5),
    n = 20, innov = rnorm(20))

  expect_type(hhj(series), "list")
  expect_type(
    suppressWarnings(
      hhj(series, sub_sample = 2, k = "one-sided")), "list"
    )
  expect_type(
    suppressWarnings(
      hhj(series, search_grid = 5, k = "bias/variance")), "list"
    )
  expect_is(hhj(series, grid_step = 4, pilot_block_length = 3), "hhj")
  expect_error(hhj(NA))
  expect_message(hhj(series, plots = FALSE))

  expect_error(hhj(series, grid_step = 3.3))
  expect_error(hhj(series, grid_step = c(3, 2, 1)))
  expect_error(hhj(series, search_grid = 3.3))
  expect_error(hhj(series, sub_sample = length(series) + 10))
  expect_error(hhj(series, k = "testing"))

})

test_that("hhj parrallelization works", {

  series <- stats::arima.sim(list(order = c(1, 0, 0), ar = 0.5),
    n = 20, innov = rnorm(20))

  cl <- parallel::makeCluster(2)

  expect_is(hhj(series, cl = cl), "hhj")

  parallel::stopCluster(cl)

})


tests/testthat/test-pwsd.R
test_that("flat-top lag window (lambda function) works", {

  t <- runif(10, min = 0.5000001, max = 1)
  t1 <- runif(5, min = 0, max = 0.5)

  # Edge Cases
  expect_equal(lambda(0), 1)
  expect_equal(lambda(0.5), 1)
  expect_equal(lambda(2), 0)
  expect_equal(lambda(NA), NA_integer_)

  # Vector Inputs
  expect_equal(lambda(t), 2 * (1 - abs(t)))
  expect_equal(lambda(t1), rep(1, length(t1)))

})

test_that("implied hypothesis test works", {

  sim <- stats::arima.sim(list(order = c(1, 0, 0), ar = 0.5),
    n = 100, innov = rnorm(100))

  n <- length(sim)

  autocorrelation <- stats::acf(
    sim,
    lag.max = 20,
    type = "correlation",
    plot = FALSE)

  # Remove lag 0 for testing
  rho_k <- autocorrelation$acf[-1]
  rho_k_critical <- 2 * sqrt(log10(n) / n)

  expect_equal(
    implied_hypothesis(rho_k = rho_k, rho_k_critical = 0, K_N = 5),
    length(rho_k)
    )

  expect_equal(
    implied_hypothesis(rho_k = rho_k, rho_k_critical = 1, K_N = 5), 1
    )

})


test_that("pwsd works", {

  sim <- stats::arima.sim(list(order = c(1, 0, 0), ar = 0.5),
    n = 100, innov = rnorm(100))

  white_noise <- rnorm(100)

  expect_type(pwsd(sim), "list")
  expect_is(pwsd(sim, round = TRUE), "pwsd")
  expect_error(pwsd(NA))
  expect_error(pwsd(sim, m_hat = 2.2))

  pwsd(white_noise, c = 4)$parameters[, "m_hat"]

  expect_equal(pwsd(white_noise, c = 4)$parameters[, "m_hat"], c("m_hat" = 1))

})


R/plot.hhj.R
#' Plot MSE Function for HHJ Algorithm
#'
#' S3 Method for objects of class 'hhj'
#'
#' @param x an object of class 'hhj'
#' @param iter a vector of \code{hhj()} iterations to plot. \code{NULL}. All
#'  iterations are plotted by default.
#'
#' @inheritDotParams base::plot
#'
#'
#' @return No return value, called for side effects
#'
#' @export
#'
#' @examples
#'
#' \donttest{
#' # Generate AR(1) time series
#' sim <- stats::arima.sim(list(order = c(1, 0, 0), ar = 0.5),
#'                         n = 500, innov = rnorm(500))
#'
#' # Generate 'hhj' class object of optimal block length for series
#' hhj <- hhj(sim, sub_sample = 10)
#'
#' ## S3 method for class 'hhj'
#' plot(hhj)
#' }
#'
plot.hhj <- function(x, iter = NULL, ...) {

  if (!inherits(x, "hhj")) {
    stop("use only with \"hhj\" objects")
  }

  p.data <- x$`MSE Data`

  if (is.null(iter)) {
    iter <- unique(p.data$Iteration)
  } else {
    stopifnot(is.numeric(iter) | is.list(iter))
  }

  # Plot all iterations
  for (i in iter) {

    # Filter on iteration
    p.data <- x$`MSE Data`[x$`MSE Data`$Iteration == i, ]

    graphics::plot.default(
      x = as.vector(p.data$BlockLength, mode = "numeric"),
      y = p.data$MSE,
      main = paste0(
        "MSE Plot for: ",
        x$Series,
        "\n",
        "Iteration: ",
        p.data$Iteration[1]
      ),
      xlab = "Block Length (l)",
      ylab = "MSE",
      col = ifelse(p.data$MSE == min(p.data$MSE), "red", "black")
    )
  }
}


R/plot.pwsd.R
#' Plot Correlogram for Politis and White Auto\eqn{-}Correlation Implied Hypothesis Test
#'
#' S3 Method for objects of class 'pwsd'
#' \emph{See} \code{?plot.acf} of the \pkg{stats} package for more customization
#'  options on the correlogram, from which \code{plot.pwsd} is based
#'
#' @param x an of object of class 'pwsd' or 'acf'
#' @param main an overall title for the plot, if no string is supplied a default
#'  title  will be populated. \emph{See} \code{\link[graphics]{title}}
#' @param ylim a numeric of length 2 giving the y-axis limits for the plot
#'
#' @inheritParams pwsd
#' @inheritDotParams base::plot
#'
#'
#' @return No return value, called for side effects
#'
#' @export
#'
#' @examples
#' # Use S3 Method
#'
#' # Generate AR(1) time series
#' sim <- stats::arima.sim(list(order = c(1, 0, 0), ar = 0.5),
#'                         n = 500, innov = rnorm(500))
#'
#' b <- pwsd(sim, round = TRUE, correlogram = FALSE)
#' plot(b)
#'
plot.pwsd <- function(x, c = NULL, main = NULL, ylim = NULL, ...) {

  # Check class
  stopifnot(any(class(x) == "pwsd"))

  # Extract ACF
  acf <- x$Acf
  names <- rownames(x$BlockLength)

  # Set c to input if not NULL
  if (is.null(c)) {
    c <- x$parameters[, "c"]
  }

  # Set significance bands
  rho_k_critical <- c * sqrt(log10(x$parameters[, "n"]) / x$parameters[, "n"])

  # Plot acf
  j <- 1
  for (i in acf) {

    # Set plot title
    if (is.null(main)) {
      main <- paste0("Correlogram for: ", names[j])
    }

    # Set y-axis limits
    if (is.null(ylim)) {
      ylim <- range(i$acf, rho_k_critical, -rho_k_critical)
    }

    # Plot ACF
    plot(i,
      ci = 0,
      xlab = "Lag (k)",
      main = main,
      ylim = ylim,
      ...)

    # Add implied significance bands
    graphics::abline(h = c(rho_k_critical, -rho_k_critical),
      col = "darkmagenta",
      lty = "dotdash")

    # Name Counter
    j <- j + 1

  }
}


R/hhj.R
#' Hall, Horowitz, and Jing (1995) "HHJ" Algorithm to Select the Optimal Block-Length
#'
#' Perform the Hall, Horowitz, and Jing (1995) "HHJ" cross-validation algorithm
#' to select the optimal block-length for a bootstrap on dependent data
#' (block-bootstrap). Dependent data such as stationary time series are suitable
#' for usage with the HHJ algorithm.
#'
#' The HHJ algorithm is computationally intensive as it relies on a
#' cross-validation process using a type of subsampling to estimate the mean
#' squared error (\eqn{MSE}) incurred by the bootstrap at various block-lengths.
#'
#' Under-the-hood, \code{hhj()} makes use of \code{\link[tseries]{tsbootstrap}},
#' \emph{see} Trapletti and Hornik (2020), to perform the moving block-bootstrap
#' (or the \emph{block-of-blocks} bootstrap by setting \code{bofb > 1}) according
#' to Kunsch (1989).
#'
#' @param series a numeric vector or time series giving the original data for
#'  which to find the optimal block-length for.
#' @param nb an integer value, number of bootstrapped series to compute.
#' @param n_iter an integer value, maximum number of iterations for the HHJ
#'  algorithm to compute.
#' @param pilot_block_length a numeric value, the block-length (\eqn{l*} in
#'  \emph{HHJ}) for which to perform initial block bootstraps.
#' @param sub_sample a numeric value, the length of each overlapping
#'  subsample, \eqn{m} in \emph{HHJ}.
#' @param k a character string, either \code{"bias/variance"},
#'  \code{"one-sided"}, or \code{"two-sided"} depending on the desired object of
#'  estimation. If the desired bootstrap statistic is bias or variance then
#'  select \code{"bias/variance"} which sets \eqn{k = 3} per HHJ. If the object
#'  of estimation is the one-sided or two-sided distribution function, then set
#'  \code{k = "one-sided"} or \code{k = "two-sided"} which sets \eqn{k = 4} and
#'  \eqn{k = 5}, respectively. For the purpose of generating symmetric confidence
#'  intervals around an unknown parameter, \code{k = "two-sided"} (the default)
#'  should be used.
#' @param bofb a numeric value, length of the basic blocks in the
#'  \emph{block-of-blocks} bootstrap, \emph{see} \code{m = } for
#'  \code{\link[tseries]{tsbootstrap}} and Kunsch (1989).
#' @param search_grid a numeric value, the range of solutions around \eqn{l*} to
#'  evaluate within the \eqn{MSE} function \emph{after} the first iteration. The
#'  first iteration will search through all the possible block-lengths unless
#'  specified in \code{grid_step = }.
#' @param grid_step a numeric value or vector of at most length 2, the number of
#'  steps to increment over the subsample block-lengths when evaluating the
#'  \eqn{MSE} function. If \code{grid_step = 1} then each block-length will be
#'  evaluated in the \eqn{MSE} function. If \code{grid_step > 1}, the \eqn{MSE}
#'  function will search over the sequence of block-lengths from \code{1} to
#'  \code{m} by \code{grid_step}. If \code{grid_step} is a vector of length 2,
#'  the first iteration will step by the first element of \code{grid_step} and
#'  subsequent iterations will step by the second element.
#' @param cl a cluster object, created by package \pkg{parallel},
#'  \pkg{doParallel}, or \pkg{snow}. If \code{NULL}, no parallelization will be
#'  used.
#' @param verbose a logical value, if set to \code{FALSE} then no interim
#'  messages are output to the console. Error messages will still be output.
#'  Default is \code{TRUE}.
#' @param plots a logical value, if set to \code{FALSE} then no interim plots
#'  are output to the console. Default is \code{TRUE}.
#'
#' @return an object of class 'hhj'
#'
#' @section References:
#'
#' Adrian Trapletti and Kurt Hornik (2020). tseries: Time Series Analysis and
#'      Computational Finance. R package version 0.10-48.
#'
#' Kunsch, H. (1989) The Jackknife and the Bootstrap for General Stationary
#'      Observations. The Annals of Statistics, 17(3), 1217-1241. Retrieved
#'      February 16, 2021, from \url{http://www.jstor.org/stable/2241719}
#'
#' Peter Hall, Joel L. Horowitz, Bing-Yi Jing, On blocking rules for the
#'      bootstrap with dependent data, Biometrika, Volume 82, Issue 3,
#'      September 1995, Pages 561-574, DOI: \doi{10.1093/biomet/82.3.561}
#'
#' @export
#'
#' @examples
#' \donttest{
#' # Generate AR(1) time series
#' sim <- stats::arima.sim(list(order = c(1, 0, 0), ar = 0.5),
#'                         n = 500, innov = rnorm(500))
#'
#' # Calculate optimal block length for series
#' hhj(sim, sub_sample = 10)
#'
#'
#' # Use parallel computing
#' library(parallel)
#'
#' # Make cluster object with 2 cores
#' cl <- makeCluster(2)
#'
#' # Calculate optimal block length for series
#' hhj(sim, cl = cl)
#' }
#'
hhj <- function(
  series,
  nb = 100L,
  n_iter = 10L,
  pilot_block_length = NULL,
  sub_sample = NULL,
  k = "two-sided",
  bofb = 1L,
  search_grid = NULL,
  grid_step = c(1L, 1L),
  cl = NULL,
  verbose = TRUE,
  plots = TRUE
  ) {

  # Check arguments
  stopifnot(class(series) %in% c("integer", "numeric", "ts"))
  stopifnot(all.equal(nb %% 1, 0))
  stopifnot(all.equal(n_iter %% 1, 0))
  stopifnot(class(grid_step) %in% c("integer", "numeric"))

  if (length(grid_step) < 2) {
    stopifnot(all.equal(grid_step %% 1, 0))
    grid_step <- c(grid_step, grid_step)
  } else if (length(grid_step) > 2) {
    stop("grid_step must be a vector of at most length 2")
  }

  if (!is.null(search_grid)) {
    stopifnot(all.equal(search_grid %% 1, 0))
  }

  # Save function call
  call <- match.call()

  # Length of whole series
  n <- length(series)

  # Set pilot block-length
  if (is.null(pilot_block_length)) {
    l_star <- round(n^(1 / 5))
  } else {
    l_star <- round(pilot_block_length)
  }

  # Print pilot message
  if (isTRUE(verbose)) {
    message(" Pilot block length is: ", l_star)
  }

  # Set estimation type (k)
  if (k == "bias/variance") {
    k <- 3
  } else if (k == "one-sided") {
    k <- 4
  } else if (k == "two-sided") {
    k <- 5
  } else {
    stop("k must be in c('bias/variance', 'one-sided', 'two-sided')")
  }

  # Check block-length of subsamples
  if (is.null(sub_sample)) {
    m <- round(n^(1 / 5) * n^(1 / k))
  } else if (sub_sample >= n) {
    stop("sub_sample must be less than series length")
  } else {
    m <- round(sub_sample)
  }

  # Initialize overlapping sub samples list
  series.list <- vector(mode = "list", length = length((n - m + 1)))

  for (j in 1:n_iter) {

    # Bootstrap variance of whole series
    boot_temp <- tseries::tsbootstrap(series,
      statistic = stats::var,
      type = "block",
      nb = nb,
      b = l_star,
      m = bofb
    )

    # Save updated variance of whole series
    v_star <- mean(boot_temp$statistic)

    # Search total grid on first iteration then +/- search_grid over next
    if (j == 1) {
      grid <- list(seq(from = 1, to = m, by = grid_step[1]))
    } else if (!is.null(search_grid)) {
      grid <- list(seq(
        from = max(1, l_m - search_grid),
        to = min(m, l_m + search_grid),
        by = grid_step[2]
      ))
    } else {
      grid <- list(seq(from = 1, to = m, by = grid_step[2]))
    }

    # Setup environment for cluster workers
    if (!is.null(cl)) {

      if (!requireNamespace("parallel", quietly = TRUE)) {
        stop("Package \"parallel\" needed for this function to work. Please install it.",
          call. = FALSE)
      }

      # Send internal MSE function parameters to each cluster
      parallel::clusterExport(cl = cl, list(
        "n", "m", "series.list", "nb", "bofb", "v_star"
        ), envir = environment())
    }

    # Minimization message
    if (j == 1 & isTRUE(verbose)) {

      # Output message to wait for gridSearch to process
      message("Performing minimization may take some time")
      message("Calculating MSE for each level in subsample: ",
        lengths(grid), " function evaluations required.")
    }

    # Create sub-blocks of length m = sub_sample
    for (i in seq_len(length.out = (n - m + 1))) {
      series.list[[i]] <- series[seq(from = i, to = (i + m - 1), by = 1)]
    }

    # Run optimization process
    if (is.null(cl)) {

      # Optimize MSE function over l in serial
      sol <- sapply(X = grid[[1]], FUN = hhjMSE,
        # Bootstrap parameters for hhjMSE
        n = n,
        m = m,
        series.list = series.list,
        nb = nb,
        bofb = bofb,
        v_star = v_star)

    } else {

      # Optimize MSE function over l in parallel
      sol <- parallel::parSapply(cl = cl, X = grid[[1]], FUN = hhjMSE,
        # Bootstrap parameters for hhjMSE
        n = n,
        m = m,
        series.list = series.list,
        nb = nb,
        bofb = bofb,
        v_star = v_star)
    }

    # Save plot data
    if (j == 1) {
      p.data <- data.frame(
        Iteration = j,
        BlockLength = round(grid[[1]] * ((n / m)^(1 / k))),
        MSE = sol
        )
    } else {
      p.data <- rbind(p.data, data.frame(
        Iteration = j,
        BlockLength = round(grid[[1]] * ((n / m)^(1 / k))),
        MSE = sol
        ))
    }

    # Plot MSE over l and color minimizing value red
    if (isTRUE(plots)) {
      plot(
        x = round(grid[[1]] * ((n / m)^(1 / k))),
        y = sol,
        main = paste0(
          "MSE Plot for: ",
          deparse(substitute(series)),
          "\n",
          "Iteration: ",
          j
        ),

        xlab = "Block Length (l)",
        ylab = "MSE",
        col = ifelse(sol == min(sol), "red", "black")
      )
    }

    # Save l that minimizes MSE of subsample blocks
    l_m <- round(which.min(sol))

    # Break if l_m converges to previous l*
    if (l_star == round((n / m)^(1 / k) * l_m)) {

      # Print convergence message
      if (isTRUE(verbose)) {
        message(" Converged at block length (l): ", round(l_star))
      }

      # Compile results list with custom class
      result <- structure(
        list(
          "Optimal Block Length" = l_star,
          "Subsample block size (m)" = m,
          "MSE Data" = p.data,
          "Iterations" = j,
          "Series" = deparse(substitute(series)),
          "Call" = call
          ),
        class = "hhj")

      # Return list of results
      return(result)
    }

    # Scale l_m back to l_n for next iteration
    l_star <- round(((n / m)^(1 / k)) * l_m)

    # Print iteration message
    if (isTRUE(verbose)) {
      message(" Chosen block length: ", l_star, "  After iteration: ", j)
    }

    # Print if final-iteration is reached w/ no convergence
    if (j == n_iter) {

      # Warning message about no convergence
      warning(
        "Block-length has not converged. Stopping at iteration limit: ", j, "\n",
        "  Result may be spurious"
        )

      # Compile results list with custom class
      result <- structure(
        list(
          "Optimal Block Length" = l_star,
          "Subsample block size (m)" = m,
          "MSE Data" = p.data,
          "Iterations" = j,
          "Series" = deparse(substitute(series)),
          "Call" = call
        ),
        class = "hhj")

      # Return list of results
      return(result)
    }

  }
}

# Helper Functions --------------------------------------------------------

# MSE function to optimize
hhjMSE <- function(l, n, m, series.list, nb, bofb, v_star) {

  # Initialize Squared Error Vector
  se <- rep(NA, (n - m + 1))

  # Bootstrap each subsample, non-parallel computation
  output <- lapply(
    X = series.list,
    FUN = tseries::tsbootstrap,
    statistic = stats::var,
    type = "block",
    nb = nb,
    b = l,
    m = bofb
  )

  for (i in seq_len(length.out = length(output))) {

    # Calculate Squared Error
    se[i] <- (mean(output[[i]]$statistic) - v_star)^2
  }

  # Calculate MSE
  return(mean(se))

}
